src/System.Linq/src/System/Linq/Concat.cs

```diff
@@ -226,182 +101,188 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
+                
+                bool reservedFirst = builder.ReserveOrAdd(_first);
+                bool reservedSecond = builder.ReserveOrAdd(_second);
 
-                if (totalCount == 0)
+                TSource[] array = builder.ToArray();
+                
+                if (reservedFirst)
                 {
-                    return Array.Empty<TSource>();
+                    Marker marker = builder.Markers.First();
+                    Debug.Assert(marker.Index == 0);
+                    EnumerableHelpers.Copy(_first, array, 0, marker.Count);
                 }
 
-                var result = new TSource[totalCount];
-
-                _first.CopyTo(result, 0);
-                _second.CopyTo(result, firstCount);
+                if (reservedSecond)
+                {
+                    Marker marker = builder.Markers.Last();
+                    EnumerableHelpers.Copy(_second, array, marker.Index, marker.Count);
+                }
 
-                return result;
+                return array;
             }
-
-            public override int GetCount(bool onlyIfCheap) => Count; // Getting the count is always cheap.
         }
 
         /// <summary>
-        /// Represents the concatenation of three or more <see cref="ICollection{TSource}"/>.
+        /// Represents the concatenation of three or more <see cref="IEnumerable{TSource}"/>.
         /// </summary>
-        /// <typeparam name="TSource">The type of the source collections.</typeparam>
-        private sealed class ConcatNCollectionIterator<TSource> : ConcatIterator<TSource>
+        /// <typeparam name="TSource">The type of the source enumerables.</typeparam>
+        /// <remarks>
+        /// To handle chains of >= 3 sources, we store each enumerable in a singly-linked list node.
+        /// Each iterator holds a reference to the head of the list, which contains the most recently
+        /// concatenated enumerable. When a new one is added, we allocate a new node that points to our
+        /// head and store it, then allocate a new iterator to wrap that node.
+        ///
+        /// This linked-list implementation means that rather than each MoveNext/Current call having to
+        /// traverse all of the previous sources, we only have to traverse all of the previous sources
+        /// once per chained enumerable.  An alternative would be to use an array to store all of the
+        /// enumerables, but this has a much better memory profile and without much additional run-time cost.
+        ///
+        /// At face value, using list nodes seems worse than chaining together the iterators directly and
+        /// letting the iterator itself hold a reference to the latest source. So why allocate a list node in
+        /// addition to each iterator? It lets the GC reclaim intermediary iterators in the chain which
+        /// have fields that are not relevant to later iterators, like <see cref="_enumerator"/> or
```

Comment:
@stephentoub There are other aspects where this is a win. We don't pay for typecasting to ICollection up front in Concat anymore, it is deferred until ToArray/ToList is called and iteration tends to be more common. `GetEnumerable` also becomes a [lot faster and simpler](https://github.com/dotnet/corefx/pull/15943/files#diff-d8b3be41edbd1e2e3c65380665467a2dR207) because the nodes are uniform type, so we don't have to typecast with O(N^2) behavior.